src/prepared_statement.c

```diff
@@ -0,0 +1,387 @@
+	ps->query_len = query_len;
+	ps->param_len = num_parameters;
+	ps->query_id = next_unique_query_id;
+	ps->use_count = 0;
+	memcpy(ps->query_with_params, pkt->query, query_len);
+	memcpy(&ps->query_with_params[query_len], pkt->parameter_types_bytes, num_parameters);
+	return ps;
+}
+
+static PgClientPreparedStatement*
+create_client_prepared_statement(char const* name, PgPreparedStatement* ps)
+{
+	size_t name_len = strlen(name) + 1;
+	PgClientPreparedStatement *client_ps = (PgClientPreparedStatement*)malloc(sizeof(PgClientPreparedStatement) + name_len);
+	memcpy(client_ps->stmt_name, name, name_len);
+	client_ps->ps = ps;
+	ps->use_count += 1;
+	return client_ps;
+}
+
+static PgServerPreparedStatement*
+create_server_prepared_statement(PgPreparedStatement* ps)
+{
+	PgServerPreparedStatement *server_ps = (PgServerPreparedStatement*)malloc(sizeof(PgServerPreparedStatement));
+	server_ps->ps = ps;
+	server_ps->query_id = ps->query_id;
+	ps->use_count += 1;
+	return server_ps;
+}
+
+static PgPreparedStatement*
+get_prepared_statement(PgParsePacket *pkt, bool* found)
+{
+	PgPreparedStatement *ps = create_prepared_statement(pkt);
+	PgPreparedStatement *old_ps = NULL;
+	HASH_FIND(hh, prepared_statements, ps->query_with_params, ps->query_len + ps->param_len, old_ps);
+	if (old_ps != NULL)
+	{
+		free(ps);
+		ps = old_ps;
+		*found = true;
+	}
+	else
+	{
+		HASH_ADD_KEYPTR(hh, prepared_statements, ps->query_with_params, ps->query_len + ps->param_len, ps);
+		ps->stmt_name_len = (uint8_t)snprintf(ps->stmt_name, sizeof ps->stmt_name, "PS_%u", ps->query_id);
+		next_unique_query_id += 1;
+		*found = false;
+	}
+	return ps;
+}
+
+static void
+extract_parse_packet(PgPreparedStatement* ps, PgParsePacket *pkt)
+{
+	pkt->query = ps->query_with_params;
+	pkt->num_parameters = ps->param_len/4;
+	pkt->parameter_types_bytes = (uint8_t *)ps->query_with_params + ps->query_len;
+}
+
+static bool
+register_prepared_statement(PgSocket *server, PgServerPreparedStatement *server_ps)
+{
+	struct PgServerPreparedStatement *current, *tmp;
+	PktBuf *buf;
+	unsigned int cached_query_count = HASH_COUNT(server->server_prepared_statements);
+
+	if (cached_query_count >= (unsigned int)cf_prepared_statement_cache_size)
+	{
+		/* Remove least recently used statement */
+		HASH_ITER(hh, server->server_prepared_statements, current, tmp) {
+			HASH_DEL(server->server_prepared_statements, current);
+			cached_query_count--;
+
+			buf = create_close_packet(current->ps->stmt_name);
+
+			if (!pktbuf_send_immediate(buf, server)) {
+				pktbuf_free(buf);
+				return false;
+			}
+			pktbuf_free(buf);
+			server->ignore_close_response = true;
+
+			slog_noise(server, "prepared statement '%s' deleted from server cache", current->ps->stmt_name);
+
+			if (--current->ps->use_count == 0)
+			{
+				HASH_DEL(prepared_statements, current->ps);
+				free(current->ps);
+			}
+			free(current);
+			break;
```

Comment:
Should only break once cached_query_count is less than cf_prepared_statement_cache_queries. During normal operation only once should indeed be enough, but when cf_prepared_statement_cache_size is lowered, this would allow us to pick up this change.

Reply:
Good catch.
Fixed